### ISSUE TITLE
Pull correct session timeout from the Context in MemcachedBackupSessionManager

### ISSUE THREAD

**Issue by ghost** [magro/memcached-session-manager#48]
_From [jhrigg...@gmail.com](https://code.google.com/u/114746121494859435020/) on May 08, 2009 16:35:26_

MemcachedBackupSessionManager does not properly set the maxInactiveInterval
for sessions (it is using the default 60 seconds rather than the
sessionTimeout value from the Context).  StandardManager can be referenced
for implementation; see setContainer(Container) and
propertyChange(PropertyChangeEvent).

_Original issue: http://code.google.com/p/memcached-session-manager/issues/detail?id=4_

**Comment by google-code-exporter** [comment]
_From [jhrigg...@gmail.com](https://code.google.com/u/114746121494859435020/) on May 08, 2009 20:40:59_

See
<a href="http://github.com/jhriggs/memcached-session-manager/commit/ba8e9655e13082e549ed4794ad1196121b6ba16e" rel="nofollow">http://github.com/jhriggs/memcached-session-manager/commit/ba8e9655e13082e549ed4794ad1196121b6ba16e</a>

**Comment by google-code-exporter** [comment]
_From [martin.grotzke](https://code.google.com/u/martin.grotzke/) on May 09, 2009 16:15:35_

Thanx for fixing this, this is merged now.

**Status:** Fixed

**Comment by cnzzr** [comment]
memcached-session-manager-1.9.1.jar with session-timeout mistake!

@marcust

**Comment by magro** [comment]
@cnzzr Can you provide s.th. reproducable, e.g. based on https://github.com/magro/memcached-session-manager/tree/master/samples?

**Comment by cnzzr** [comment]
tomcat 6.0.45
JDK 1.6
windows 2008 R2 SP1 this problem exits on windows 2003 server.
@magro 

**web.xml config**

  <session-config>
    <session-timeout>60</session-timeout>
  </session-config>

**use Sticky**
app context config:
<Manager className="de.javakaffee.web.msm.MemcachedBackupSessionManager"
    memcachedNodes="n2:localhost:11211,zny:10.219.8.8:11211"
    failoverNodes="n2"
    lockingMode="uriPattern:/login|/logout"
    requestUriIgnorePattern=".*\.(ico|png|gif|jpg|css|js|gzjs|gzcss)$"
    transcoderFactoryClass="de.javakaffee.web.msm.serializer.javolution.JavolutionTranscoderFactory"
    />

**view tomcat session with probe 2.4**
![20160218](https://cloud.githubusercontent.com/assets/239363/13132786/dbdd2560-d630-11e5-8163-7a29fca3faa7.png)

**Comment by magro** [comment]
Ok, I need to reproduce this.

**Comment by magro** [comment]
@cnzzr I can't reproduce this with the [sample app](https://github.com/magro/memcached-session-manager/tree/master/samples). It's already [setting the session-timeout to 3 minutes](https://github.com/magro/memcached-session-manager/blob/master/samples/simpleservlet/src/main/webapp/WEB-INF/web.xml#L31) and when I print the `session.getMaxInactiveInterval()` in the [CounterServlet](https://github.com/magro/memcached-session-manager/blob/master/samples/simpleservlet/src/main/java/de/javakaffee/web/msm/sample/CounterServlet.java) it also gives me 180 (seconds) - as expected. Btw, I'm running it according to the README with `$ mvn tomcat7:run -pl :simpleservlet -am` and then get [http://localhost:9090/counter](http://localhost:9090/counter).

Can you please provide a reproduce sample?

**Comment by cnzzr** [comment]
hello @magro 
i just test memcached-session-manager 1.9.1 with tomcat 7 session.getMaxInactiveInterval() value is right.

this problem  just exits with tomcat6

my test app just a jsp file.
```
<%
out.println( "seconds=" + session.getMaxInactiveInterval() + "<br/>");
%>
```

**Comment by magro** [comment]
Strange. I just changed the samples to use tomcat6, please see and follow instructions in commit 8e5bcfa0 to check this (you have to checkout this commit or the branch "issue-48-sessionTimeout-tomcat6").

**Comment by cnzzr** [comment]
tomcat6.0.37  with jdk6
 + session.getMaxInactiveInterval(): 180

@magro

**Comment by magro** [comment]
@cnzzr So it seems to be working as expected, right? If it's not working in your application it seems to be application specific, you might try to continually simplify your app (and/or tomcat/webapp setup) to see when it starts to work.

**Comment by cnzzr** [comment]
diff tomcat version
tomcat 6.0.45
mvn tomcat is 6.0.37

i have try msm with tomcat 6.0.45 with clear app just one jsp page
app1 without msm show 1800
app2 with msm sho 30

**Comment by magro** [comment]
@cnzzr I just updated the [issue-48-sessionTimeout-tomcat6 branch](https://github.com/magro/memcached-session-manager/tree/issue-48-sessionTimeout-tomcat6) to use 6.0.45 - this allows to reproduce the issue (while with 6.0.44 it's working as expected).

Btw, in #287 a quite similar issue is reported for the latest tomcat 7.0.68 - maybe both issues share the same root cause.

**Comment by magro** [comment]
@cnzzr @kimmob I think I found s.th.: tomcat guys decided to deprecate the [Manager.maxInactiveInterval](https://tomcat.apache.org/tomcat-7.0-doc/api/org/apache/catalina/Manager.html#getMaxInactiveInterval()) in a new patch version and always just use the `Context.getSessionTimeout()` (default session timeout). This explains the wrong/changed behaviour. Other Manager implementations from tomcat also still invoke `Manager.setMaxInactiveInterval`, so it's not obvious what has to be changed.

Maybe you could investigate further why Manager.maxInactiveInterval was deprecated and what should be used instead to use the session timeout from the webapp/web.xml?

**Comment by magro** [comment]
Related:
https://github.com/apache/tomcat/commit/8bc6fa5140ea23e97922ff07b345b92450da3618
https://github.com/apache/tomcat/commit/7e8dcf3cf03fd8956ab9ae1535a6787949ae5c44

**Comment by magro** [comment]
In tomcat 6 and tomcat 7 Manager.getMaxInactiveInterval is broken, while in tomcat 8 it's correct:
https://github.com/apache/tomcat60/blob/trunk/java/org/apache/catalina/session/ManagerBase.java#L596
https://github.com/apache/tomcat70/blob/trunk/java/org/apache/catalina/session/ManagerBase.java#L426
https://github.com/apache/tomcat80/blob/trunk/java/org/apache/catalina/session/ManagerBase.java#L427

Also had to tell this tomcat's twitter account ;-)
https://twitter.com/martin_grotzke/status/702110946035163136
https://twitter.com/martin_grotzke/status/702112668837142528

**Comment by magro** [comment]
@cnzzr The issue-48-sessionTimeout-tomcat6 now works for me, I'll release msm with the changes. If you can confirm that it works this would be great.

**Comment by magro** [comment]
D'oh, there are failing tests, and with the minutes based Context.sessionTimeout tests are running forever. Needs investigation / another solution.

**Comment by magro** [comment]
@cnzzr I just released 1.9.2 with the fix.
